File: go/day2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func solve_day2_part1(input string) (string, error) {
	horizontal := 0
	depth := 0

	for _, line := range strings.Split(input, "\n") {
		line := strings.TrimSpace(line)
		if len(line) == 0 {
			continue
		}
		value, err := strconv.Atoi(strings.Split(line, " ")[1])
		if err != nil {
			return "", err
		}
		if strings.Index(line, "forward") == 0 {
			horizontal += value
		} else if strings.Index(line, "down") == 0 {
			depth += value
		} else if strings.Index(line, "up") == 0 {
			depth -= value
		} else {
			return "", fmt.Errorf("Unexpected command: %d", value)
		}
	}
	return strconv.FormatInt(int64(horizontal*depth), 10), nil
}

func solve_day2_part2(input string) (string, error) {
	horizontal := 0
	depth := 0
	aim := 0

	for _, line := range strings.Split(input, "\n") {
		line := strings.TrimSpace(line)
		if len(line) == 0 {
			continue
		}
		value, err := strconv.Atoi(strings.Split(line, " ")[1])
		if err != nil {
			return "", err
		}
		if strings.Index(line, "forward") == 0 {
			horizontal += value
			depth += aim * value
		} else if strings.Index(line, "down") == 0 {
			aim += value
		} else if strings.Index(line, "up") == 0 {
			aim -= value
		} else {
			return "", fmt.Errorf("Unexpected command: %d", value)
		}
	}
	return strconv.FormatInt(int64(horizontal*depth), 10), nil
}

File: rust/src/day2.rs
pub fn solve_part_1(input: &str) -> anyhow::Result<impl std::fmt::Display> {
    let (mut horizontal, mut depth) = (0isize, 0isize);

    for line in input
        .split('\n')
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
    {
        let value: isize = line.split(" ").last().ok_or(anyhow!("No value"))?.parse()?;
        if line.starts_with("forward") {
            horizontal += value;
        } else if line.starts_with("down") {
            depth += value;
        } else if line.starts_with("up") {
            depth -= value;
        } else {
            bail!("Unexpected command: {}", value)
        }
    }

    Ok(horizontal * depth)
}

pub fn solve_part_2(input: &str) -> anyhow::Result<impl std::fmt::Display> {
    let (mut horizontal, mut depth, mut aim) = (0isize, 0isize, 0isize);

    for line in input
        .split('\n')
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
    {
        let value: isize = line.split(" ").last().ok_or(anyhow!("No value"))?.parse()?;
        if line.starts_with("forward") {
            horizontal += value;
            depth += aim * value;
        } else if line.starts_with("down") {
            aim += value;
        } else if line.starts_with("up") {
            aim -= value;
        } else {
            bail!("Unexpected command: {}", value)
        }
    }

    Ok(horizontal * depth)
}
